app/src/play/java/com/celzero/bravedns/StoreAppUpdater.kt

```diff
@@ -0,0 +1,96 @@
+        AppUpdateManagerFactory.create(context)
+    }
+
+    override fun completeUpdate() {
+        appUpdateManager.completeUpdate()
+    }
+
+    override fun unregisterListener(listener: AppUpdater.InstallStateListener) {
+        listenerMapping.remove(listener)?.also {
+            appUpdateManager.unregisterListener(it)
+        }
+    }
+
+    override fun checkForAppUpdate(isUserInitiated: Boolean, activity: Activity, listener: AppUpdater.InstallStateListener) {
+        Log.d(LOG_TAG, "Beginning update check.")
+        val playListener = InstallStateUpdatedListener { state ->
+            val mappedStatus = when (state.installStatus()) {
+                InstallStatus.DOWNLOADED -> AppUpdater.InstallStatus.DOWNLOADED
+                InstallStatus.CANCELED -> AppUpdater.InstallStatus.CANCELED
+                InstallStatus.DOWNLOADING -> AppUpdater.InstallStatus.DOWNLOADING
+                InstallStatus.FAILED -> AppUpdater.InstallStatus.FAILED
+                InstallStatus.INSTALLED -> AppUpdater.InstallStatus.INSTALLED
+                InstallStatus.INSTALLING -> AppUpdater.InstallStatus.INSTALLING
+                InstallStatus.PENDING -> AppUpdater.InstallStatus.PENDING
+                else -> AppUpdater.InstallStatus.UNKNOWN
+            }
+            listener.onStateUpdate(AppUpdater.InstallState(mappedStatus))
+        }
+        listenerMapping[listener] = playListener
+        appUpdateManager.registerListener(playListener)
+
+        appUpdateManager.appUpdateInfo.addOnSuccessListener { appUpdateInfo ->
+            if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE && appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.FLEXIBLE)) {
+                Log.i(LOG_TAG, "Update available, starting flexible update")
+                try {
+                    appUpdateManager.startUpdateFlowForResult(appUpdateInfo, AppUpdateType.FLEXIBLE, activity, 1)
+                } catch (e: IntentSender.SendIntentException) {
+                    unregisterListener(listener)
+                    Log.e(LOG_TAG, "SendIntentException: ${e.message} ", e)
+                }
+            } else if (appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE && appUpdateInfo.isUpdateTypeAllowed(AppUpdateType.IMMEDIATE)) {
+                Log.i(LOG_TAG, "Update available, starting immediate update")
+                try {
+                    appUpdateManager.startUpdateFlowForResult(appUpdateInfo, AppUpdateType.IMMEDIATE, activity, 1)
+                } catch (e: IntentSender.SendIntentException) {
+                    unregisterListener(listener)
+                    Log.e(LOG_TAG, "SendIntentException: ${e.message} ", e)
+                }
+            } else {
+                unregisterListener(listener)
+                Log.e(LOG_TAG, "no update or updating not allowed.")
+                if (isUserInitiated) {
+                    listener.onUpdateCheckFailed()
```

Comment:
I copied this over from the old logic, but you might want to have another check here. Currently this'll tell the user the update check failed when he is up-to-date or in-app-updates aren't allowed currently (because the quota has already been reached)

Reply:
Can you please add that as a `TODO` in the code comment?

Reply:
It's only a few if-statements, I can add them. It was more of a "do you want to have it this way"-question

Reply:
> Currently this'll tell the user the update check failed when it is up-to-date or in-app-updates aren't allowed currently (because the quota has already been reached).

> It was more of a "do you want to have it this way"-question

In that case and if I understood you right, from what you said, it'd be right to show the right thing here that's less ambiguous and misleading, that is, either "the app is up-to-date" or "to try again later" (in case of quota overflow)?

Reply:
Done